src/EditorFeatures/Core/Tagging/AsynchronousTaggerProvider.Tagger.cs

```diff
@@ -0,0 +1,257 @@
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Tagging;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.Tagging
+{
+    internal partial class AsynchronousTaggerProvider<TTag>
+    {
+        private sealed partial class Tagger : IAccurateTagger<TTag>, IDisposable
+        {
+            private const int MaxNumberOfRequestedSpans = 100;
+
+            #region Fields that can be accessed from either thread
+
+            private readonly ITextBuffer _subjectBuffer;
+
+            private readonly TagSource _tagSource;
+
+            #endregion
+
+            #region Fields that can only be accessed from the foreground thread
+
+            /// <summary>
+            /// The batch change notifier that we use to throttle update to the UI.
+            /// </summary>
+            private readonly BatchChangeNotifier _batchChangeNotifier;
+
+            #endregion
+
+            public event EventHandler<SnapshotSpanEventArgs> TagsChanged;
+
+            public Tagger(
+                IAsynchronousOperationListener listener,
+                IForegroundNotificationService notificationService,
+                TagSource tagSource,
+                ITextBuffer subjectBuffer)
+            {
+                Contract.ThrowIfNull(subjectBuffer);
+
+                _subjectBuffer = subjectBuffer;
+                _batchChangeNotifier = new BatchChangeNotifier(subjectBuffer, listener, notificationService, ReportChangedSpan);
+
+                _tagSource = tagSource;
+
+                _tagSource.OnTaggerAdded(this);
+                _tagSource.TagsChangedForBuffer += OnTagsChangedForBuffer;
+                _tagSource.Paused += OnPaused;
+                _tagSource.Resumed += OnResumed;
+            }
+
+            public void Dispose()
+            {
+                _tagSource.Resumed -= OnResumed;
+                _tagSource.Paused -= OnPaused;
+                _tagSource.TagsChangedForBuffer -= OnTagsChangedForBuffer;
+                _tagSource.OnTaggerDisposed(this);
+            }
+
+            private void ReportChangedSpan(SnapshotSpan changeSpan)
+            {
+                var tagsChanged = TagsChanged;
+                if (tagsChanged != null)
+                {
+                    tagsChanged(this, new SnapshotSpanEventArgs(changeSpan));
+                }
+            }
+
+            private void OnPaused(object sender, EventArgs e)
+            {
+                _batchChangeNotifier.Pause();
+            }
+
+            private void OnResumed(object sender, EventArgs e)
+            {
+                _batchChangeNotifier.Resume();
+            }
+
+            private void OnTagsChangedForBuffer(ICollection<KeyValuePair<ITextBuffer, NormalizedSnapshotSpanCollection>> changes)
+            {
+                // Note: This operation is uncancellable. Once we've been notified here, our cached tags
```

Comment:
I suppose not.  But it shoudn't be a huge deal.  when we finally process the result, we'll see that we don't have anyone attached to our change notification and we'll simply no-op.